vmm/src/vm.rs

```diff
@@ -564,19 +620,38 @@ impl DeviceManager {
 
             virtio_pci_device.assign_msix(msi_cb);
         } else {
-            let irqfd = EventFd::new(EFD_NONBLOCK).map_err(DeviceManagerError::EventFd)?;
-
             let irq_num = allocator
                 .allocate_irq()
                 .ok_or(DeviceManagerError::AllocateIrq)?;
-            vm_fd
-                .register_irqfd(irqfd.as_raw_fd(), irq_num)
-                .map_err(DeviceManagerError::Irq)?;
 
-            let irq_cb = Arc::new(
+            let irq_cb = if let Some(ioapic) = interrupt_info.ioapic {
+                let ioapic_clone = ioapic.clone();
+                Box::new(move |_p: InterruptParameters| {
+                    ioapic_clone
+                        .lock()
+                        .unwrap()
+                        .service_irq(irq_num as usize)
+                        .map_err(|e| {
+                            std::io::Error::new(
+                                std::io::ErrorKind::Other,
+                                format!("failed to inject IRQ #{}: {:?}", irq_num, e),
+                            )
+                        })
+                }) as InterruptDelivery
+            } else {
+                let irqfd = EventFd::new(EFD_NONBLOCK).map_err(DeviceManagerError::EventFd)?;
+                vm_fd
+                    .register_irqfd(irqfd.as_raw_fd(), irq_num)
+                    .map_err(DeviceManagerError::Irq)?;
+
                 Box::new(move |_p: InterruptParameters| irqfd.write(1)) as InterruptDelivery
+            };
```

Comment:
Just a longer term comment: We have `DeviceInterrupt` and `InterruptDelivery` callbacks, for essentially doing the same thing: Delivering an interrupt to the guest. I think we should be looking at converging those 2 into one single trait.
This would make the VMM code clearer and easier to follow.

Reply:
Agreed. To be specific, I'm looking for something more generic Event (or EventChannel) trait and implementations, just need some abstract around Linux-specific EventFd. It's not only limited to host-to-guest event, but also guest-to-host event, and even inter-communication events between vmm and vhost-user-backend, etc.
```rust
trait Event {
 pub notify(value)
}

struct IrqEvent {
    eventfd: Option<EvendFd>;
}
impl Event for IrqEvent {
    pub new(eventfd: EventFd) {
    }
    pub notify(value: Option) {
        eventfd.write(1);
    }
}
```

Your latest version is very close to this and only the naming could be more generic.